providers/dynamodb/dynamodb.go

```diff
@@ -0,0 +1,106 @@
+// DynamoDB storage abstraction layer
+package dynamodb
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+)
+
+type DynamoDB struct {
+	Table string
+	Conn  *dynamodb.DynamoDB
+}
+
+// New creates an instance of the DynamoDB struct for us.
+func New(access string, secret string, region string, tableName string) (DynamoDB, error) {
+	creds := credentials.NewStaticCredentials(access, secret, "")
+
+	cfg := aws.Config{
+		Region: aws.String(region),
+	}
+
+	cfg.WithCredentials(creds)
+
+	ddb := DynamoDB{
+		Table: tableName,
+		Conn:  dynamodb.New(session.New(), &cfg),
+	}
+
+	return ddb, nil
+}
+
+// Put overwrites or creates as needed a new file based on the key.
+func (ddb DynamoDB) Put(key string, data []byte) error {
+
+	if len(data) == 0 {
+		return errors.New("Can't put empty values")
+	}
+
+	params := &dynamodb.PutItemInput{
+		TableName: aws.String(ddb.Table),
+		Item: map[string]*dynamodb.AttributeValue{
+			"key": &dynamodb.AttributeValue{
+				S: aws.String(key),
+			},
+			"value": &dynamodb.AttributeValue{
+				B: data,
+			},
+		},
+	}
+	_, err := ddb.Conn.PutItem(params)
+	return err
+}
+
+// Get attempts to grab data from dynamodb
+func (ddb DynamoDB) Get(key string) []byte {
+	params := &dynamodb.GetItemInput{
+		TableName: aws.String(ddb.Table),
+		Key: map[string]*dynamodb.AttributeValue{
+			"key": {
+				S: aws.String(key),
+			},
+		},
+	}
+
+	results, err := ddb.Conn.GetItem(params)
+
+	if err != nil {
+		return []byte{}
+	}
+
+	if val, ok := results.Item["value"]; ok {
+		return val.B
+	}
+	return []byte{}
```

Comment:
You're right, our package is highly inconsistent and should be resolved:

 - [Bolt](https://github.com/kyani-inc/storage/blob/master/providers%2Fbolt%2Fbolt.go#L32-L41) - Handles nil values
 - [Redis](https://github.com/garyburd/redigo/blob/master/redis%2Freply.go#L175) - They are returning nil as a value which makes our Redis abstraction inconsistent because we don't catch that, just return it as nil.
- [In Memory](https://github.com/kyani-inc/storage/blob/master/providers%2Flocal%2Flocal.go#L12) - it's a `map[string][]byte` which an uninitialized `[]byte` is nil
 - [Folder](https://golang.org/src/bytes/buffer.go?s=12902:12936#L396) - Uses `ioutil.ReadAll` which piggy backs off of `bytes.NewBuffer()` which it itself returns nil (default value of the pointer)
- [Memcached)(https://github.com/bradfitz/gomemcache/blob/master/memcache/memcache.go#L303) - They are checking for nil on their side which we bubble back
- [S3](https://github.com/kyani-inc/storage/blob/master/providers%2Fs3%2Fs3.go#L68-L83) - We return an empty `[]byte` here if something goes wrong with S3.

So looking at all of our packages S3 and DynamoDB return empty []bytes. Redis, In Memory, Folder, Bolt and Memcached all have the possibility to return nil.

If we want to adopt a return empty []byte strategy so be it. Document it and I'll be happy. But we'll need to update all the functions to catch nil and assert it to an empty byte to follow the contract.